[Serializer] Fix accessor-method false positive detection.

whsv26 opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.3 <!-- see below -->
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | Fix #... <!-- prefix each issue number with "Fix #", no need to create an issue if none exist, explain below instead -->
| License       | MIT
| Doc PR        | 


Fix for cases when methods like ```hashCode``` are treated as accessor methods by ObjectNormalizer.

<!--
Replace this notice by a short README for your feature/bugfix. This will help people
understand your PR and can be used as a start for the documentation.

Additionally (see https://symfony.com/releases):
 - Always add tests and ensure they pass.
 - Never break backward compatibility (see https://symfony.com/bc).
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against branch 5.x.
 - Changelog entry should follow https://symfony.com/doc/current/contributing/code/conventions.html#writing-a-changelog-entry
-->

whsv26 committed:
[Serializer] Fix accessor-method false positive detection.

stof commented on src/Symfony/Component/Serializer/Normalizer/ObjectNormalizer.php:
```diff
@@ -190,4 +190,11 @@ protected function getAllowedAttributes($classOrObject, array $context, bool $at
         return $allowedAttributes;
     }
+
+    private function strStartsWith(string $str, string $prefix): bool
```
the method name does not describe at all what is does

whsv26 replied:
@stof is it better now?

whsv26 committed:
[Serializer] Fix accessor-method false positive detection.

Rename strStartsWith to more meaningful name.
Remove code duplication for issers and getters/hassers.

whsv26 committed:
[Serializer] Fix accessor-method false positive detection.

Coding standard patch.

whsv26 committed:
[Serializer] Fix passing null to parameter php 8.1 deprecation.

whsv26 committed:
[Serializer] Fix passing null to parameter php 8.1 deprecation.

whsv26 commented:
I'm not sure what I can do about 7.4, high-deps test.

OskarStark commented on src/Symfony/Component/Serializer/Tests/Normalizer/ObjectNormalizerTest.php:
```diff
@@ -302,6 +302,20 @@ public function testConstructorWithUnknownObjectTypeHintDenormalize()
+    {
+        $obj = new DummyWithAccessorLikes();
+
+        $this->assertEquals(
```
```suggestion
        $this->assertSame(
```

whsv26 committed:
[Serializer] More strict accessor distinction assert.

nicolas-grekas commented on src/Symfony/Component/Serializer/Normalizer/ObjectNormalizer.php:
```diff
@@ -190,4 +188,37 @@ protected function getAllowedAttributes($classOrObject, array $context, bool $at
+     * Extract accessor prefix from method name if possible.
+     *
+     * @psalm-pure
+     * @psalm-return (value-of<self::ACCESSOR_PREFIXES>)|null
```
please use regular annotations everywhere

nicolas-grekas commented on src/Symfony/Component/Serializer/Normalizer/ObjectNormalizer.php:
```diff
@@ -190,4 +188,37 @@ protected function getAllowedAttributes($classOrObject, array $context, bool $at
+     * @psalm-pure
+     * @psalm-param (value-of<self::ACCESSOR_PREFIXES>) $accessorPrefix
+     */
+    private function extractAttributeName(string $accessorMethod, string $accessorPrefix): ?string
```
please remove this method and inline the implementation

nicolas-grekas commented:
Can you please rebase/retarget on 4.4 if the bug is also found there?
